ur_robot_driver/src/ros/hardware_interface.cpp

```diff
@@ -628,6 +648,38 @@ void HardwareInterface::publishPose()
   }
 }
 
+void HardwareInterface::extractRobotStatus()
+{
+  using namespace rtde_interface;
+
+  robot_status_resource_.mode =
+      robot_status_bits_[UrRtdeRobotStatusBits::IS_TEACH_BUTTON_PRESSED] ? RobotMode::MANUAL : RobotMode::AUTO;
+
+  robot_status_resource_.e_stopped =
+      safety_status_bits_[UrRtdeSafetyStatusBits::IS_EMERGENCY_STOPPED] ? TriState::TRUE : TriState::FALSE;
+
+  // Note that this is true as soon as the drives are powered,
+  // even if the breakes are still closed
+  // which is in slight contrast to the comments in the
+  // message definition
+  robot_status_resource_.drives_powered =
+      robot_status_bits_[UrRtdeRobotStatusBits::IS_POWER_ON] ? TriState::TRUE : TriState::FALSE;
+
+  robot_status_resource_.motion_possible =
+      robot_mode_ == ur_dashboard_msgs::RobotMode::RUNNING ? TriState::TRUE : TriState::FALSE;
+
+  // I found no way to reliably get information if the robot is moving
+  robot_status_resource_.in_motion = TriState::UNKNOWN;
+
+  // note that e-stop is handled by a seperate variable
+  robot_status_resource_.in_error =
+      safety_status_bits_[UrRtdeSafetyStatusBits::IS_PROTECTIVE_STOPPED] ? TriState::TRUE : TriState::FALSE;
```

Comment:
One thing I believe we could include in this PR: `in_error` should be the union of anything which makes the controller go into an error state. Is `IS_EMERGENCY_STOPPED` also a cause of `IS_PROTECTIVE_STOPPED`? If not, then it would need to be `||` here I believe.

Any other causes we could add to `in_error`?

Reply:
Sure, I can add this. Unfortunately I cannot test emergency stops, since I don't have a real robot here, just ursim. I don't know the behaviour of these bits, therefore I cannot tell which bits provide additional information regarding the error state.

Reply:
I think all of these should be used for that:
```
  IS_PROTECTIVE_STOPPED = 2,
  IS_ROBOT_EMERGENCY_STOPPED = 6,
  IS_EMERGENCY_STOPPED = 7,
  IS_VIOLATION = 8,
  IS_FAULT = 9,
  IS_STOPPED_DUE_TO_SAFETY = 10
```

Reply:
I could make a short trip to the lab these days to test the behavior.